Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceDomGenerator.cpp

```diff
@@ -221,9 +222,12 @@ namespace AzToolsFramework
             PrefabDomPath domPathToContainerEntity(pathToContainerEntity.c_str());
 
             // Retrieves the container entity DOM of the root template DOM.
-            const PrefabDom& rootTemplateDom = m_prefabSystemComponentInterface->FindTemplateDom((*rootInstancePtr).GetTemplateId());
             PrefabDom containerEntityDom;
-            containerEntityDom.CopyFrom(*domPathToContainerEntity.Get(rootTemplateDom), instanceDom.GetAllocator());
+            {
+                // rootTemplateDom is only valid in this scope.
+                const PrefabDom& rootTemplateDom = m_prefabSystemComponentInterface->FindTemplateDom((*rootInstancePtr).GetTemplateId());
+                containerEntityDom.CopyFrom(*domPathToContainerEntity.Get(rootTemplateDom), instanceDom.GetAllocator());
```

Comment:
Here too

Reply:
Current code:
```cpp
// Retrieves the container entity DOM of the root template DOM.
PrefabDom containerEntityDom;
const PrefabDom& rootTemplateDom = m_prefabSystemComponentInterface->FindTemplateDom((*rootInstancePtr).GetTemplateId());
containerEntityDom.CopyFrom(*domPathToContainerEntity.Get(rootTemplateDom), 
    instanceDom.GetAllocator());

// Sets the container entity DOM in the given instance DOM.
const AZStd::string containerEntityName = AZStd::string::format("/%s", PrefabDomUtils::ContainerEntityName);
PrefabDomPath containerEntityPath(containerEntityName.c_str());
containerEntityPath.Set(instanceDom, containerEntityDom, instanceDom.GetAllocator());
```
This is almost the same pattern like we did in [preemptive update](https://github.com/o3de/o3de/blob/f93742e99582b2de00e7d3c4bf390b7e4e0668cb/Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabUndo.cpp) for instance DOM cache.

Looks like the way it could work fine is because we are using `instanceDom.GetAllocator()` for containerEntityDom. If I change it to `containerEntityDom.GetAllocator()`, it crashed because the last line Set doesn't make Value 'owned' by instanceDom.

Reason is `rapidjson::Pointer::Set()` will directly use the allocator defined in the Value if it is a Document. So in this case, containerEntityDom is a document and providing the allocator in Set() seems redundant. As of which allocator it will use, it looks a bit tricky.

More details: https://rapidjson.org/md_doc_pointer.html (Resolving Document and Value)

Reply:
Alternatively, a better solution would be not make a copy at all. Set() internally calls CopyFrom already using the document's allocator OR the one provided.

```cpp
const PrefabDom& rootTemplateDom = m_prefabSystemComponentInterface->FindTemplateDom((*rootInstancePtr).GetTemplateId());
const PrefabDomValue* containerEntityDom = domPathToContainerEntity.Get(rootTemplateDom);

// Sets the container entity DOM in the given instance DOM.
const AZStd::string containerEntityName = AZStd::string::format("/%s", PrefabDomUtils::ContainerEntityName);
PrefabDomPath containerEntityPath(containerEntityName.c_str());
containerEntityPath.Set(instanceDom, *containerEntityDom);
// OR containerEntityPath.Set(instanceDom, *containerEntityDom, instanceDom.GetAllocator());  // same
```

Reply:
In that case, shouldn't the `containerEntityPath.Set` call on line 236 below AZStd::move the `containerEntityDom`

Reply:
yeah. It will be using move version of Set: [rapidjson/pointer.h#L771](https://github.com/Tencent/rapidjson/blob/master/include/rapidjson/pointer.h#L771)

We can do this so it makes a copy (by ourselves) then move it to the document.
```cpp
containerEntityPath.Set(instanceDom, 
    PrefabDomValue(*containerEntityDom, instanceDom.GetAllocator()).Move());
```